feat: deposit flow refactor

srene opened the pull request:
This PR is aimed at improving deposit flow, removing fixed waiting time to reduce latency and retrying in case failed validation for retryable cases.

changes:
* finalization is checked on relayer before submitting for validation on deposit detected instead of fixed waiting time after each loop.
* if deposit is not final enough, deposit is queued with delay depending on pending confirmations.
* validator checks finalization and returns error based on that
* depending on error type for failed validations relayer retries deposit using exp backoff.

srene committed:
deposit fail reason

srene committed:
fix

srene committed:
finality and reorg merge

srene committed:
fix

srene committed:
reason change

srene committed:
queue and retry after failed

srene committed:
retryng before validation

srene committed:
comment

srene committed:
revert test files changes

srene committed:
minor

srene committed:
minor

srene committed:
minor

srene committed:
minor

srene committed:
fix compiling issue

srene committed:
config files update

srene committed:
Merge branch 'main-dym' into srene/deposit-flow-refactor

srene committed:
Merge branch 'main-dym' into srene/deposit-flow-refactor

srene committed:
dont requeue non retryable error

srene committed:
comment

srene committed:
restore file

danwt committed:
fmt

danwt commented on dymension/libs/kaspa/lib/core/src/finality/finality.rs:
```diff
@@ -15,13 +15,21 @@ pub fn is_mature(daa_score_block: u64, daa_score_virtual: u64, network_id: Netwo
+/// Result of finality check with detailed status information
+#[derive(Debug, Clone)]
+pub struct FinalityStatus {
+    pub is_final: bool,
```
is_final shouldnt be a field because its computable

danwt commented on dymension/libs/kaspa/lib/relayer/src/deposit/deposit.rs:
```diff
@@ -1,14 +1,102 @@
+    ProcessingError(eyre::Error),
+}
+
+impl std::fmt::Display for DepositError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            DepositError::NotFinalEnough {
+                confirmations,
+                required_confirmations,
+                retry_after_secs,
+            } => {
+                write!(
+                    f,
+                    "Deposit not final enough: {}/{} confirmations. Retry in {:.1}s",
+                    confirmations, required_confirmations, retry_after_secs
+                )
+            }
+            DepositError::ProcessingError(err) => {
+                write!(f, "Processing error: {}", err)
+            }
+        }
+    }
+}
```
cant this be derived?

danwt commented on dymension/libs/kaspa/lib/relayer/src/deposit/deposit.rs:
```diff
@@ -1,14 +1,102 @@
+    deposit: &Deposit,
+    rest_client: &HttpClient,
+) -> Result<Option<DepositFXG>, DepositError> {
+    // Check if the deposit is safe against reorg first
```
useless comment

danwt commented on dymension/libs/kaspa/lib/relayer/src/deposit/deposit.rs:
```diff
@@ -1,14 +1,102 @@
+        // we assume 10 confirmations per second, so retry after 0.1 seconds per confirmation needed
+        let retry_after_secs = pending_confirmations as f64 * 0.1;
+
+        warn!(
+            "Deposit {} is not yet safe against reorg. Confirmations: {}/{}. Will retry in {:.1}s",
+            deposit.id,
+            finality_status.confirmations,
+            finality_status.required_confirmations,
+            retry_after_secs
+        );
```
dont use warn logs they are just annoying

danwt commented on dymension/libs/kaspa/lib/relayer/src/deposit/deposit.rs:
```diff
@@ -1,14 +1,102 @@
+        });
+    }
+
+    info!(
+        "Deposit {} is safe against reorg with {} confirmations",
+        deposit.id, finality_status.confirmations
+    );
```
dont interpolate data in logs, put it at the end

danwt commented on dymension/libs/kaspa/lib/relayer/src/deposit/deposit.rs:
```diff
@@ -1,14 +1,102 @@
+        deposit.id, finality_status.confirmations
+    );
+
+    // decode payload into Hyperlane message
```
uesless comment

danwt commented on rust/main/chains/dymension-kaspa/src/providers/validators.rs:
```diff
@@ -77,11 +77,28 @@ impl ValidatorsClient {
-                        h, e
-                    );
-                    Err(e)
+                    let error_str = e.to_string();
+                    if error_str.contains("TransactionRejected") {
+                        error!(
+                            "Dymension, transaction rejected, validator: {:?}, error: {:?}",
+                            h, e
+                        );
+                        // This is non-retryable - mark as permanent failure
+                        Err(e)
+                    } else if error_str.contains("ServiceUnavailable") {
+                        error!(
+                            "Dymension, service unavailable, validator: {:?}, error: {:?}",
+                            h, e
+                        );
+                        // This is retryable with longer backoff
+                        Err(e)
+                    } else {
+                        error!(
+                            "Dymension, got deposit sig response Err, validator: {:?}, error: {:?}",
+                            h, e
+                        );
+                        Err(e)
+                    }
```
cant we do something type safe instead of string contains? need to figure out how that works across http boundary

danwt commented on rust/main/chains/dymension-kaspa/src/providers/validators.rs:
```diff
@@ -223,11 +240,28 @@ pub async fn request_validate_new_deposits(
+
+        // Create more specific errors based on HTTP status code
+        let error = match status {
+            StatusCode::ACCEPTED => {
+                // 202: Deposit not final (retryable with backoff)
+                eyre::eyre!("DepositNotFinal: {}", error_msg)
```
accepted means not final, how does that make sense?

danwt commented on rust/main/chains/dymension-kaspa/src/providers/validators.rs:
```diff
@@ -223,11 +240,28 @@ pub async fn request_validate_new_deposits(
+            StatusCode::ACCEPTED => {
+                // 202: Deposit not final (retryable with backoff)
+                eyre::eyre!("DepositNotFinal: {}", error_msg)
+            }
+            StatusCode::UNPROCESSABLE_ENTITY => {
+                // 422: Transaction rejected (non-retryable)
+                eyre::eyre!("TransactionRejected: {}", error_msg)
+            }
+            StatusCode::SERVICE_UNAVAILABLE => {
+                // 503: Service unavailable (retryable)
+                eyre::eyre!("ServiceUnavailable: {}", error_msg)
+            }
+            _ => {
+                // Default error format
+                eyre::eyre!("ValidationFailed: {} - {}", status, error_msg)
+            }
+        };
+
```
lets by typesafe here, not string based

danwt commented on rust/main/chains/dymension-kaspa/src/validator_server.rs:
```diff
@@ -42,13 +42,28 @@ impl IntoResponse for AppError {
-            format!("Validation failed: {}", err_msg),
-        )
-            .into_response()
+        // Differentiate between error types for different HTTP status codes
+        let (status_code, response_body) = if err_msg.contains("not safe against reorg") {
+            // Use 202 Accepted for non-final deposits (retryable)
+            (
+                StatusCode::ACCEPTED,
+                format!("Deposit not final: {}", err_msg),
+            )
+        } else if err_msg.contains("Hub is not bootstrapped") {
+            // Use 503 Service Unavailable for infrastructure issues (retryable)
+            (
+                StatusCode::SERVICE_UNAVAILABLE,
+                format!("Service unavailable: {}", err_msg),
+            )
+        } else {
+            // Default to 500 for other validation errors
+            (
+                StatusCode::INTERNAL_SERVER_ERROR,
+                format!("Validation failed: {}", err_msg),
+            )
+        };
+
+        (status_code, response_body).into_response()
```
should follow google https://google.aip.dev/193#handling_errors aip 193 especially regarding status codes

danwt commented on rust/main/hyperlane-base/src/kas_hack/deposit_operation.rs:
```diff
@@ -0,0 +1,100 @@
+    pub fn is_empty(&self) -> bool {
+        self.operations.is_empty()
+    }
+}
```
looks like a lot of custom stuff which is similar to
https://github.com/dymensionxyz/hyperlane-monorepo/blob/f8d9af997298501f00c32a6a68d4b755a190a9d5/rust/main/hyperlane-core/src/traits/pending_operation.rs#L1-L426
https://github.com/dymensionxyz/hyperlane-monorepo/blob/f8d9af997298501f00c32a6a68d4b755a190a9d5/rust/main/agents/relayer/src/msg/pending_message.rs#L1-L1412

danwt commented on rust/main/hyperlane-base/src/kas_hack/logic_loop.rs:
```diff
@@ -129,44 +137,113 @@ where
+
+        // Process ready operations
+        while let Some(operation) = queue.pop_ready() {
+            drop(queue); // Release lock before processing
```
you drop the whole queue on every loop?

danwt commented on rust/main/hyperlane-base/src/kas_hack/logic_loop.rs:
```diff
@@ -129,44 +137,113 @@ where
+        }
+    }
+
+    /// Process a single deposit operation, with retry logic on failure
+    async fn process_deposit_operation(&self, mut operation: DepositOperation) {
```
this is called in two places, once from the queueu and once directly
all messages should just have a linear flow: observed -> in queue -> process attempt (and maybe back on queue)

danwt commented on rust/main/hyperlane-base/src/kas_hack/logic_loop.rs:
```diff
@@ -241,6 +318,12 @@ where
+    fn is_retryable_chain_error(&self, error: &ChainCommunicationError) -> bool {
+        let error_str = format!("{:?}", error);
+        !error_str.contains("TransactionRejected")
+    }
```
avoid str contains

danwt reviewed (commented):
We should also lift out the new hardcoded stuff, and make them configurable

danwt committed:
fix: surgical improvements to Kaspa deposit flow (#237) (claude)

* fix: surgical improvements to Kaspa deposit flow

- Add type-safe error handling with KaspaDepositError enum
- Extract hardcoded values to KaspaDepositConfig
- Fix logging practices (no warn, proper structured logging)
- Prepare foundation for PendingOperation integration
- Remove unnecessary code comments
- Use proper error type matching where possible

Addresses review comments on PR #221

* fix: make the implementation fully working end-to-end

- Remove incomplete PendingDeposit implementation (was not building)
- Properly integrate error handling with relayer's DepositError
- Fix all logging to use structured fields (no string interpolation)
- Add proper error conversion and categorization
- Ensure the code compiles and builds successfully
- Use KaspaDepositError consistently throughout

This is now a fully working implementation that:
- Builds without errors
- Has proper type-safe error handling
- Uses consistent structured logging
- Properly converts between error types
- Maintains backward compatibility

* fix: properly integrate configuration from provider level

- Move KaspaDepositConfig to dymension-kaspa crate where it belongs
- Add to RelayerStuff struct for proper initialization
- Expose via KaspaProvider::kaspa_deposit_config() method
- Update Foo constructor to get config from provider
- Support environment variables for runtime configuration
- Make conf module public for proper access

Addresses review comment about configuration integration

* fix: remove env parsing and string matching

- Remove from_env() method, configuration now comes through constructor
- Add kaspa_deposit_config parameter to ConnectionConf::new
- Parse config values in build_kaspa_connection_conf from settings
- Remove string matching in error conversion (anti-pattern)
- Simplify KaspaDepositError enum to only essential variants
- All errors now retryable in simplified version

This addresses review feedback about proper configuration flow and
avoiding string comparisons for error handling

* remove unused, fmt

* finality func

* fix: remove unused max_retries field from KaspaDepositConfig

The max_retries field was defined but never actually used to enforce retry limits. Removing it to avoid confusion.

* fixup
